pkg/pmem-csi-driver/nodeserver.go

```diff
@@ -88,57 +101,92 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
-		return &csi.NodePublishVolumeResponse{}, nil
-	}
+	srcPath := req.GetStagingTargetPath()
+	ephemeral := false
 
-	if err := os.Mkdir(targetPath, os.FileMode(0755)); err != nil {
-		// Kubernetes is violating the CSI spec and creates the
-		// directory for us
-		// (https://github.com/kubernetes/kubernetes/issues/75535). We
-		// allow that by ignoring the "already exists" error.
-		if !os.IsExist(err) {
-			return nil, status.Error(codes.Internal, "make target dir: "+err.Error())
+	// Kubernetes v1.16+ would request ephemeral volumes via VolumeContext
+	val, ok := req.GetVolumeContext()[volumeContextEphemeral]
+	if ok {
+		b, err := strconv.ParseBool(val)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
 		}
+		ephemeral = b
+	} else {
+		// If met all below conditions we still treat as a ephemeral volume request:
+		// 1) No Volume found with given volume id
+		// 2) No provisioner info found in VolumeContext "storage.kubernetes.io/csiProvisionerIdentity"
+		// 3) No StagingPath in the request
+		device, _ := ns.cs.dm.GetDevice(req.VolumeId)
+		_, ok := req.GetVolumeContext()[volumeProvisionerIdentity]
+		ephemeral = device == nil && !ok && len(srcPath) == 0
 	}
 
-	readOnly := req.GetReadonly()
-	attrib := req.GetVolumeContext()
-	mountFlags := req.GetVolumeCapability().GetMount().GetMountFlags()
+	var mountOptions []string
 
-	klog.V(3).Infof("NodePublishVolume: id:%s targetpath:%v stagingtargetpath:%v readonly:%v attributes:%v mountflags:%v",
-		req.GetVolumeId(), targetPath, stagingtargetPath, readOnly, attrib, mountFlags)
+	if ephemeral {
+		val, ok := req.GetVolumeContext()["size"]
+		if !ok {
+			return nil, status.Error(codes.InvalidArgument, "size not found in volume context")
+		}
+
+		size, err := resource.ParseQuantity(val)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("Failed to parse given size '%s': %s", val, err.Error()))
+		}
+
+		// Create new device
+		params := req.GetVolumeContext()
+		params[pmemParameterKeyPersistencyModel] = string(pmemPersistencyModelEphemeral)
+		createReq := &csi.CreateVolumeRequest{
+			VolumeCapabilities: []*csi.VolumeCapability{req.VolumeCapability},
+			Name:               req.VolumeId,
+			CapacityRange:      &csi.CapacityRange{RequiredBytes: size.Value()},
+			Parameters:         params,
+		}
 
-	options := []string{"bind"}
-	if readOnly {
-		options = append(options, "ro")
+		createResp, err := ns.cs.CreateVolume(ctx, createReq)
+		if err != nil {
+			return nil, status.Error(codes.Internal, fmt.Sprintf("Failed to create new ephemeral volume: %s", err.Error()))
+		}
+
+		device, err := ns.cs.dm.GetDevice(createResp.Volume.VolumeId)
+		if err != nil {
+			return nil, status.Error(codes.Internal, fmt.Sprintf("Could not retrieve newly created volume with id '%s': %s", createResp.Volume.VolumeId, err.Error()))
+		}
+
+		// Create filesystem
+		err = ns.provisionDevice(device, req.GetVolumeCapability().GetMount().GetFsType())
```

Comment:
We don't have tests which use a non-default filesystem, do we? Let's not make this a blocker for merging this PR, but keep it in mind: https://github.com/intel/pmem-csi/issues/431

Reply:
> We don't have tests which use a non-default filesystem, do we?

I am not sure about ephemral volumes, But at least for persistent volumes, Volumes test-suite covers this. Here is the needed change in our test code: https://github.com/avalluri/pmem-CSI/commit/bd357fbd757e12f32cadcc5d915aa0b2e6a18264

Reply:
I was thinking of ephemeral volumes here.
_

Reply:
Yes, current ephemeral tests are quite limited and not supported to choose the filesystem type for volume.